DataViews: add support for `NOT IN` operator in filter

oandregal opened the pull request:
Part of https://github.com/WordPress/gutenberg/issues/55083
Related to https://github.com/WordPress/gutenberg/issues/55100

## What?

Adds support for `NOT IN` operator in the `Author` filter while leaving `Status` as it is (only supports `IN` operator).

https://github.com/WordPress/gutenberg/assets/583546/6e022bd5-b725-4423-8fe9-37dc0c8385a1

## Why?

It's part of the interactions we want to achieve.

## How?

- Fields API:
  - by default, any field of `type: ENUMERATION` gets the filter with both operator (`IN`, and `NOT IN`).
  - this can be configurable by providing the specific ones in the `filterBy.operators` property (see how the `Author` field uses it to only enable the `IN` operator).
- Filter component: add support for both operators.

## Testing Instructions

- Enable the "admin views" Gutenberg experiment and visit "Appearance > Editor > Pages > Manage all pages".
- Interact with the filters.

oandregal committed:
Add support for not in filter

oandregal committed:
Field API: allow declaring which operators a field supports

oandregal commented on packages/edit-site/src/components/dataviews/in-filter.js:
```diff
@@ -31,49 +83,111 @@ export default ( { filter, view, onChangeView } ) => {
-									},
-								],
-							} ) )
+			<WithSeparators>
```
This changes from:

```jsx
<DropdownMenu>
  <DropdownMenuCheckboxItem />
  <DropdownMenuCheckboxItem />
</DropdownMenu>
```

to:

```jsx
<DropdownMenu>
  <WithSeparators>
    <DropdownMenuGroup>
      <DropdownMenuCheckboxItem />
      <DropdownMenuCheckboxItem />
    </DropdownMenuGroup>
    <DropdownSubMenu>
      <DropdownMenuCheckboxItem />
      <DropdownMenuCheckboxItem />
    </DropdownSubMenu>
  </WithSeparators>
</DropdownMenu>
```

Essentially adds a new `DropdownSubmenu` subtree. The rest are indentation changes that result of adding the `WithSeparators` component.

youknowriad commented:
Design wise should we do something more like notion where you don't select a "value" from the "add filter" dropdown, instead, you just select the field, and then you get a modal/popover where you can do several things:

 - Choose an operator
 - Choose one or multiple values.

I think this modal will scale well to all operators and type of fields... Some fields for instance might not have a fixed list of values and you want to type something instead.

https://github.com/WordPress/gutenberg/assets/272444/ef6c2e51-330a-4d30-a535-26b28344b24a

jameskoster commented:
@youknowriad I think that might go beyond the scope of what's capable in the new DropdownMenu component and require the use of an actual modal. Cc @ciampo for confirmation.

That's not to say we should do it, but it would require a little design exploration. If we wanted to use DropdownMenu something like this might work (with finessed wording)?

<img width="797" alt="Screenshot 2023-11-23 at 15 13 48" src="https://github.com/WordPress/gutenberg/assets/846565/78ec1744-c783-4710-86dd-7f9228cad163">

youknowriad commented:
I think the fact that we started with filters that consist of just item selection is maybe leading us to a direction that doesn't accommodate all the needs. We're trying to force some semantics to the dropdown menu that are not well suited for it. We want a form for each filter basically and a form can be shown in a dropdown or a modal, both work but not a dropdown menu.

jameskoster commented:
That's a fair point, but it is quite well optimised for the majority of use cases – it would be a shame to make those feel more complicated unnecessarily. 

Another option would be to not include operator options in these basic filters, and offer those options in a dedicated "Advanced filter" UI instead. Notion offers something similar:

<img width="557" alt="Screenshot 2023-11-23 at 15 34 04" src="https://github.com/WordPress/gutenberg/assets/846565/9cb36eb6-4e9f-417b-bafd-2adf5c0db8e5">

This is especially comprehensive as it even allows operations between filters.

ciampo commented:
>  If we wanted to use DropdownMenu something like this might work

Using nested menus with radio options could work

oandregal commented:
I've been working on refactoring a bit the existing components to be able to add this feature all at once in all of them. I'd like to land https://github.com/WordPress/gutenberg/pull/56514 before this PR.

jameskoster commented:
I made a quick mockup for the flow Riad proposed, where the filter UIs are forms in popovers rather than DropdownMenus. 

<img width="1314" alt="Screenshot 2023-11-27 at 12 25 46" src="https://github.com/WordPress/gutenberg/assets/846565/4f66f8f3-817e-4e0e-99da-7780148e5d33">

The main difference in design is that we lose the ability to add filters as part of the flyout, and that since the new UI is a form we should probably use form elements (checkboxes and radios). I think both concessions are okay given the benefits – more flexibility in terms of what we can place in the popovers.

There's a question about where to put focus when adding a new filter, I would assume on the first element in the popover but cc @andrewhayward for any thoughts on that. Edit: Also worth double-checking whether these UIs being forms rather than menus would impact whether or not we need a dedicated 'Apply' button.

---

I appreciate the mockup above doesn't include an operator UI. I'll work on that if we decide to refactor in this direction.

oandregal committed:
Fix rebase

oandregal committed:
Fix: do not change operator

oandregal committed:
Do not register filter if it does not support any operator

oandregal commented:
> I've been working on refactoring a bit the existing components to be able to add this feature all at once in all of them. I'd like to land https://github.com/WordPress/gutenberg/pull/56514 before this PR.

I'm no longer waiting for 56514 to be merged first. This one is ready and I welcome reviews.

andrewhayward commented:
I know the `menu` foundation upon which filters are built is part of an ongoing conversation, so I wouldn't consider extending it here to be a blocker, particularly.

That being said, there are a few things I'll flag, some of which are more important than others.


### Consistency

<p align="center">
  <img src="https://github.com/WordPress/gutenberg/assets/159848/14cae785-7339-46ab-ab68-087f9b1f42f7" alt="A dropdown menu from a trigger button labelled 'Author', highlighting check marks in mixed positions, and a large gap where icons might be." width="45%" valign="top">
  <img src="https://github.com/WordPress/gutenberg/assets/159848/cd254c54-8861-4d56-aab0-a084ff49badd" alt="A dropdown menu from a trigger button labelled 'Author is admin', highlighting check marks consistently on the left." width="45%" valign="top">
</p>

Might feel like a minor point, but the "selected" check marks are inconsistent across menus. The menu that opens from table headers has them on the right, but then on the left when in the Settings submenu. The menu that opens from a filter button however has them consistently on the left, even for the same filter options.

Note also that there's a gap on the right of the filter options in the table menu, where icons or other markers might appear.


### Focus management

Not an issue introduced here, but noting nevertheless. Sometimes, when selecting an item in a table header menu, focus is lost to the winds. This seems to be a latency issue, because I can only reproduce it with fresh filter settings. If I filter or sort by a previously requested value the data is populated immediately and focus remains on the header element.


### Semantics

Different platforms have different considerations when it comes to interdependent menu options, so it's not a huge deal if we use check marks to denote selected items in a mutually exclusive group.

<img src="https://github.com/WordPress/gutenberg/assets/159848/d9eb75ec-3b8b-48ef-9682-a38a8732b952" alt="A menu with two mutually exclusive options, circled in red." width="40%" align="left">

However, the underlying semantics should be correct, regardless of the visuals. The options in the main filter menu, for example, are (currently) single-select, but have a role of `menuitemcheckbox`. This should be `menuitemradio`, to indicate that only one item in the group can be picked.

The same is also true of options in the Settings submenu, and (although unrelated to this PR) the sorting options, which have a role of `menuitem` and no `aria-checked` attribute. <br clear="left">

ciampo commented:
> Might feel like a minor point, but the "selected" check marks are inconsistent across menus

General inconsistencies between dropdownmenus are likely caused by the fact that some menus are still using the older version of the component — that should be fixed soon, so probably we shouldn't worry about it in this PR.

oandregal committed:
FilterSummary: improve semantics

oandregal commented:
@andrewhayward I pushed some changes to fix the following things:

- Consistency for filters: all instances of filters (columns, top) show the check mark to the left. Other menus (sorting, etc.) I left as they were – it seems a left mark would not work for all of them given they have icons, but it's a separate concern from this PR anyway.
- Semantics: all instances of filters use the `role="menuitemradio"`. I've prepared a PR at [#56676](https://github.com/WordPress/gutenberg/pull/56676) looking at other dataview's menus.
- Focus: this is not introduced here, so it can be investigated separately.

jameskoster commented:
Agree with @ciampo on the dropdown, let's focus on making the new component bullet-proof in separate issues / PRs.

@oandregal this is working well. The words are a little tricky, especially as we'll need to account for any/all operators down the road too. Here's a mockup to try, but I'd welcome more feedback:

<img width="1718" alt="Screenshot 2023-11-30 at 15 26 17" src="https://github.com/WordPress/gutenberg/assets/846565/e7f8992e-4fc8-4f36-a4d8-459757366377">

Naturally in this PR we should strip that back to:

<img width="1293" alt="Screenshot 2023-11-30 at 15 26 29" src="https://github.com/WordPress/gutenberg/assets/846565/3a26d051-d544-4324-9327-923a6a6fcced">

I noticed WooCommerce use the term "Conditions" in their product filter widget. I don't love it, but I'm assuming it has stood the test of time for them.

I'll finish up by saying these details are easy to tweak after the fact if needed, and don't need to hold up merging this PR imo.

andrewhayward commented:
> Semantics: all instances of filters use the `role="menuitemradio"`. I've prepared a PR at [#56676](https://github.com/WordPress/gutenberg/pull/56676) looking at other dataview's menus.

Thanks. Don't know if this is something to be fixed here or in #56676, but every `menuitemradio` also needs a corresponding [`aria-checked`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-checked) value too (either `true` or `false`).

oandregal committed:
Add aria-checked

oandregal committed:
Use Conditions instead of Settings

oandregal committed:
Use Is/Is not

oandregal commented:
Fixed the `aria-checked` https://github.com/WordPress/gutenberg/pull/56479/commits/250babaac48562048eba9ec8e76c3645172ab98e and updated the wording as suggested https://github.com/WordPress/gutenberg/pull/56479/commits/250babaac48562048eba9ec8e76c3645172ab98e and https://github.com/WordPress/gutenberg/pull/56479/commits/7ca13ce3838ac7e2afca37555b634832f29a9605


https://github.com/WordPress/gutenberg/assets/583546/5e182efe-2d7c-4c89-8f0f-97c684744d1a

oandregal committed:
Show condition is submenu trigger

oandregal commented:
Also displayed the condition being applied in the submenu:

<img width="460" alt="Captura de ecrã 2023-11-30, às 17 38 39" src="https://github.com/WordPress/gutenberg/assets/583546/e5975413-ea61-4401-b7d7-2de0f25807aa">

https://github.com/WordPress/gutenberg/assets/583546/6e022bd5-b725-4423-8fe9-37dc0c8385a1

youknowriad commented on packages/edit-site/src/components/dataviews/filter-summary.js:
```diff
@@ -139,7 +139,16 @@ export default function FilterSummary( { filter, view, onChangeView } ) {
+										{ filterInView.operator === OPERATOR_IN
+											? __( 'Is' )
+											: __( 'Is not' ) }
+										<Icon icon={ chevronRightSmall } />{ ' ' }
```
Should we check for isRTL in all these chevronRight usage?

andrewhayward replied:
Ideally this would be handled at the component level, as in RTL environments (almost) everything should be reversed, with the suffix on the left, facing the other way.

This should just work in [the newer menu component](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-dropdownmenu-v2-ariakit--docs).

youknowriad replied:
I know that we have automatic inversions for CSS but so far, for icons we had to do it manually, so we should be consistent: either have the components handle this everywhere or not, because otherwise it will confuse developers

andrewhayward replied:
Sorry, I wasn't clear – "everything should be reversed" within the context of menu items. The newer menu component doesn't need a chevron icon passed through as a suffix at all; that bit is automatic if the the menu item has a child menu.

Otherwise, yes, [RTL icons aren't always just flipped versions of their LTR equivalents](https://developer.apple.com/design/human-interface-guidelines/right-to-left#Interface-icons), and we should leave that decision to developers.

<p align="center">
  <img src="https://github.com/WordPress/gutenberg/assets/159848/63beee3b-fb15-449e-b0b9-fdc9e7c3472b" alt="A silhouette of a speaker pointing right with a backslash on top of it." width="100">
  <img src="https://github.com/WordPress/gutenberg/assets/159848/bdb8673d-7493-4b6d-8862-a9491d062ae3" alt="A silhouette of a speaker pointing left with a backslash on top of it." width="100">
</p>

ciampo replied:
The ariakit-based version of the new `DropdownMenu` component already handles flipping the chevron icon in RTL environments (see [storybook example](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-dropdownmenu-v2-ariakit--with-submenu&globals=direction:rtl)), so I'd say no need to deal with it in this PR

andrewhayward replied:
Mostly unrelated, but as a side-note, it would be great if the icon system was a bit smarter, and could handle bi-directionality implicitly. So rather than developers having to explicitly think about context...

```js
if ( isRTL() ) {
```

<img src="https://github.com/WordPress/gutenberg/assets/159848/bced35eb-9425-44bb-8e9c-8bb5fa2a5014" alt="A pilcrow next to a left-pointing arrow." align="left" height="48px">

```jsx
<Icon icon={ icons.formatRtl } ... />
```

```js
} else {
```

<img src="https://github.com/WordPress/gutenberg/assets/159848/3748483d-4e8b-4f15-88c0-bc7edc35a008" alt="A pilcrow next to a right-pointing arrow." align="left" height="48px">

```jsx
<Icon icon={ icons.formatLtr } ... />
```

```js
}
```

...they could just do something like this...

<img src="https://github.com/WordPress/gutenberg/assets/159848/bced35eb-9425-44bb-8e9c-8bb5fa2a5014" alt="A pilcrow next to a left-pointing arrow." align="left" height="48px">

```jsx
<Icon icon={ icons.format } ... /> {/* RTL */}
```

<img src="https://github.com/WordPress/gutenberg/assets/159848/3748483d-4e8b-4f15-88c0-bc7edc35a008" alt="A pilcrow next to a right-pointing arrow." align="left" height="48px">

```jsx
<Icon icon={ icons.format } ... /> {/* LTR */}
```

`</brain-dump>`

jameskoster replied:
Would something like this be too rudimentary?

```
[dir=ltr] {
    --icon-direction: 1;
}

[dir=rtl] {
    --icon-direction: -1;
}

.icon {
    transform: scaleX(var(--icon-direction));
}
```

andrewhayward replied:
> Would something like this be too rudimentary?

Yes, because it doesn't always work.

<p align="center">
  <img src="https://github.com/WordPress/gutenberg/assets/159848/3748483d-4e8b-4f15-88c0-bc7edc35a008" alt="A pilcrow next to a right-pointing arrow." height="96px">
  <img src="https://github.com/WordPress/gutenberg/assets/159848/bced35eb-9425-44bb-8e9c-8bb5fa2a5014" alt="A pilcrow next to a left-pointing arrow." height="96px">
  <img src="https://github.com/WordPress/gutenberg/assets/159848/03b54142-0280-4f26-bc09-680416998c1d" alt="A left-pointing arrow next to a horizontally flipped pilcrow." height="96px">
</p>

We could be a bit smarter in the SVG markup though, if we wanted to keep it out of the JS...

```svg
<SVG xmlns="http://www.w3.org/2000/svg" viewBox="-2 -2 24 24">
  <Path data-dir="ltr" d="..." />
  <Path data-dir="rtl" d="..." />
</SVG>
```

```css
[dir=rtl] .icon [data-dir=ltr],
[dir=ltr] .icon [data-dir=trl] {
  display: none;
}
```

oandregal committed:
Use right or left depending on RTL

jameskoster reviewed (approved):
Happy to approve this for design. The implementation is good based on the component used, and it works well. The wording is easily tweaked later based on wider feedback.

ciampo commented:
> Thanks. Don't know if this is something to be fixed here or in #56676, but every `menuitemradio` also needs a corresponding [`aria-checked`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-checked) value too (either `true` or `false`).
> ...
> Fixed the `aria-checked` [250baba](https://github.com/WordPress/gutenberg/commit/250babaac48562048eba9ec8e76c3645172ab98e) and updated the wording as suggested [250baba](https://github.com/WordPress/gutenberg/commit/250babaac48562048eba9ec8e76c3645172ab98e) and [7ca13ce](https://github.com/WordPress/gutenberg/commit/7ca13ce3838ac7e2afca37555b634832f29a9605)

As also commented in https://github.com/WordPress/gutenberg/pull/56676#pullrequestreview-1758415785, the new `DropdownMenu` implementation offers specific components for implementing menu items — we should be using those, instead of manually tweaking `DropdownMenuItem` with extra attributes (see [this comment](https://github.com/WordPress/gutenberg/pull/56676#pullrequestreview-1758415785) for more details on how to perform the refactor)

oandregal committed:
Fix rebase

oandregal committed:
Revert "Use right or left depending on RTL"

This reverts commit 24e5ef3c261c3c10f14735b7fde747b3ee382e87.

oandregal committed:
Fix rebase

oandregal committed:
Fix AddFilter: ignore operator to decide which filters to list

oandregal commented:
It seems that I've addressed all feedback here, so I'm merging this one on Monday if no one else brings up anything.

oandregal committed:
Fix rebase

mcsf commented on packages/dataviews/src/filters.js:
```diff
@@ -4,7 +4,17 @@
-import { ENUMERATION_TYPE, OPERATOR_IN } from './constants';
+import { ENUMERATION_TYPE, OPERATOR_IN, OPERATOR_NOT_IN } from './constants';
+
+const operatorsFromField = ( field ) => {
```
Wondering if a better name might be `sanitizeOperators`.

oandregal replied:
https://github.com/WordPress/gutenberg/pull/57050

mcsf commented on packages/dataviews/src/filter-summary.js:
```diff
@@ -6,20 +6,72 @@ import {
+	) {
+		return sprintf(
+			/* translators: 1: Filter name. 2: Filter value. e.g.: "Author is Admin". */
+			__( '%1$s is %2$s' ),
```
I wonder how [more synthetic languages](https://en.wikipedia.org/wiki/Inflection) are handled in this sort of UI (e.g. Notion). I'm not discouraging this bit of string interpolation, just pointing out that we should see how others have done this, so that later we are sure that this UI is localisable.

oandregal replied:
At some point, the proposal was `Field: value`. Though it didn't scale well when you have more operators (in, not in, starts with, etc.).

This is how notion does it (suffers from the same issue):

https://github.com/WordPress/gutenberg/assets/583546/49045409-6d9f-4f30-a0d6-b8313b0aa37f

This is linear (same issue):

https://github.com/WordPress/gutenberg/assets/583546/95a3bdbe-6185-40ab-a24e-0d11992fcebe

The same in other platforms (github, etc.).

I find synthetic languages difficult for this use case because the individual parts (field name, value) are translated separately, and you have no context what they are when translating this string as a whole. Happy to look at examples to improve it.

oandregal replied:
This is German:

https://github.com/WordPress/gutenberg/assets/583546/c9945f47-ef49-4b67-b15c-9b86e4c6a031

The gender does affect other parts of the sentence, though the general structure is the same – I'd think what we have here is fine.

mcsf replied:
Thanks for having a look! Yeah, I think it's doable, and at worst there might be a handful of languages in which the UI reads slightly awkwardly but remains understandable.

mcsf commented on packages/dataviews/src/view-table.js:
```diff
@@ -70,15 +70,49 @@ function HeaderMenu( { dataView, header } ) {
-			elements: header.column.columnDef.elements || [],
-		};
+		let columnOperators = header.column.columnDef.filterBy?.operators;
+		if ( ! columnOperators || ! Array.isArray( columnOperators ) ) {
+			columnOperators = [ OPERATOR_IN, OPERATOR_NOT_IN ];
+		}
+		const operators = columnOperators.filter( ( operator ) =>
+			[ OPERATOR_IN, OPERATOR_NOT_IN ].includes( operator )
+		);
+		if ( operators.length >= 0 ) {
```
Just noticing that this is the same pattern of validation as seen in `./filters.js`

* * *

Also, is this condition right? It will always evaluate to true, as it's the length of an array.

```js
 operators.length >= 0
```

oandregal replied:
This is a bug, thanks for catching! Fix at https://github.com/WordPress/gutenberg/pull/57048

oandregal replied:
On the topic of code reuse: there's an opportunity to reuse some code between the filters.js and the view-table.js. I gave it a try at https://github.com/WordPress/gutenberg/pull/56514 though it wasn't merged. I'm happy to pair on this and take a look at how it could be improved.

mcsf replied:
Don't worry, I think premature consolidation is a danger. I was just pointing out the pattern for later. :)

oandregal replied:
Prompted by this conversation, I've taken a look and consolidated little details so all the filter implementations are mostly lookalike (but still maintaining the little UI/interaction differences for each one) https://github.com/WordPress/gutenberg/pull/57059

It's not as ambitious as 56514 was, but I've found some bugs with the column filters by doing this exercise, so it's worth having anyway.

mcsf reviewed (commented):
I left some notes, but nice work here!
